go/service/gregor.go

```diff
@@ -1016,36 +1043,88 @@ func (g *gregorHandler) auth(ctx context.Context, cli rpc.GenericClient) (err er
+	url, err := url.Parse(g.G().Env.GetGregorURI())
+	if err != nil {
+		g.Debug(ctx, "isReachable: failed to parse server uri, exiting: %s", err.Error())
+		return false
+	}
+
+	// If we currently think we are online, then make sure
+	conn, err := net.DialTimeout("tcp", url.Host, timeout)
+	if conn != nil {
+		conn.Close()
+		return true
+	}
+	if err != nil {
+		g.Debug(ctx, "isReachable: error: terminating connection: %s", err.Error())
+		g.Shutdown()
+		if err := g.Connect(g.uri); err != nil {
+			g.Debug(ctx, "isReachable: error connecting: %s", err.Error())
+		}
+		return false
+	}
+
+	return true
+}
+
 func (g *gregorHandler) pingLoop() {
 
+	ctx := context.Background()
 	id, _ := libkb.RandBytes(4)
 	duration := g.G().Env.GetGregorPingInterval()
 	timeout := g.G().Env.GetGregorPingTimeout()
-	g.Debug(context.Background(), "ping loop: starting up: id: %x duration: %v timeout: %v", id, duration, timeout)
-	defer g.Debug(context.Background(), "ping loop: id: %x terminating", id)
+	url, err := url.Parse(g.G().Env.GetGregorURI())
+	if err != nil {
+		g.Debug(ctx, "ping loop: failed to parse server uri, exiting: %s", err.Error())
+		return
+	}
+
+	g.Debug(ctx, "ping loop: starting up: id: %x duration: %v timeout: %v url: %s",
+		id, duration, timeout, url.Host)
+	defer g.Debug(ctx, "ping loop: id: %x terminating", id)
 
 	for {
+		ctx, shutdownCancel := context.WithCancel(context.Background())
 		select {
 		case <-g.G().Clock().After(duration):
 			var err error
-			ctx := context.Background()
-
-			if g.IsConnected() {
-				// If we are connected, subject the ping call to a fairly
-				// aggressive timeout so our chat stuff can be responsive
-				// to changes in connectivity
-				var cancel context.CancelFunc
-				ctx, cancel = context.WithTimeout(ctx, timeout)
-				_, err = gregor1.IncomingClient{Cli: g.pingCli}.Ping(ctx)
-				cancel()
-			} else {
-				// If we are not connected, we don't want to timeout anything
-				// Just hook into the normal reconnect chan stuff in the RPC
-				// library
-				g.Debug(ctx, "ping loop: id: %x normal ping, not connected", id)
-				_, err = gregor1.IncomingClient{Cli: g.pingCli}.Ping(ctx)
-			}
 
+			doneCh := make(chan error)
+			go func(ctx context.Context) {
+				if g.IsConnected() {
+					// If we are connected, subject the ping call to a fairly
+					// aggressive timeout so our chat stuff can be responsive
+					// to changes in connectivity
+					var timeoutCancel context.CancelFunc
+					var timeoutCtx context.Context
+					timeoutCtx, timeoutCancel = context.WithTimeout(ctx, timeout)
+					_, err = gregor1.IncomingClient{Cli: g.pingCli}.Ping(timeoutCtx)
+					timeoutCancel()
+				} else {
+					// If we are not connected, we don't want to timeout anything
+					// Just hook into the normal reconnect chan stuff in the RPC
+					// library
+					g.Debug(ctx, "ping loop: id: %x normal ping, not connected", id)
+					_, err = gregor1.IncomingClient{Cli: g.pingCli}.Ping(ctx)
```

Comment:
ok cool, might want to add something in comment about "uses the default rpc timeout"